syz-verifier/main.go

```diff
@@ -0,0 +1,329 @@
+func (srv *RPCServer) NewExchange(a *rpctype.NewExchangeArgs, r *rpctype.NewExchangeRes) error {
+	srv.mu.Lock()
+	defer srv.mu.Unlock()
+
+	if a.Info.Calls != nil {
+		res := &verf.Result{
+			ProgIdx: a.ProgIdx,
+			RnName:  a.Name,
+			Hanged:  a.Hanged,
+			Info:    a.Info,
+		}
+		if resReady := srv.newResult(res); resReady {
+			srv.sendForVerification(a.ProgIdx)
+		}
+	}
+
+	prog, pi := srv.newProgram(a.Name)
+	r.RPCProg = rpctype.RPCProg{Prog: prog, ProgIdx: pi}
+	return nil
+}
+
+// newProgram is called when a Runner request the next program to execute
+func (srv *RPCServer) newProgram(rname string) ([]byte, int) {
+	if len(srv.progs[rname]) == 0 {
+		// There are no programs that haven't been sent to the Runner yet so we need to generate a new one.
+		srv.progIdx++
+		vrf := srv.vrf
+		// TODO: decide what the number of system calls should be set to
+		prog := vrf.Target.Generate(vrf.Rnd, 2, vrf.ChoiceTable)
+		srv.serializedProgs[srv.progIdx] = prog.Serialize()
+		srv.results[srv.progIdx] = make([]*verf.Result, 0)
+		srv.left[srv.progIdx] = make([]string, 0)
+		// For each alive Runner, record the fact that there is another program to send.
+		for rn := range srv.progs {
+			srv.progs[rn] = append(srv.progs[rn], srv.progIdx)
+		}
+		for rn := range srv.runners {
+			srv.left[srv.progIdx] = append(srv.left[srv.progIdx], rn)
+		}
+	}
+
+	pi := srv.progs[rname][0]
+	srv.progs[rname] = srv.progs[rname][1:]
+	return srv.serializedProgs[pi], pi
+}
+
+// newResult is called when a Runner sends a new Result. It returns true if all Results from the corresponding programs have been received and they can be sent for verification. Otherwise, it returns false.
+func (srv *RPCServer) newResult(res *verf.Result) bool {
+	pi := res.ProgIdx
+	rname := res.RnName
+	srv.results[pi] = append(srv.results[pi], res)
+	rnsLeft := srv.left[pi]
+
+	for idx, rn := range rnsLeft {
+		if rn == rname {
+			// Find and remove the Runner that has just sent results
+			srv.left[pi] = append(rnsLeft[:idx], rnsLeft[idx+1:]...)
+			break
+		}
+	}
+
+	if len(srv.runners) > 1 && len(srv.left[pi]) == 0 {
+		// The Runner was the only alive Runner that hasn't sent results for the program and there is at least one other Runner that previously sent results. Therefore, the results can be sent for verification.
+		return true
+	}
+	return false
+}
+
+func (srv *RPCServer) sendForVerification(pi int) {
+	results, sprog := srv.results[pi], srv.serializedProgs[pi]
+	go srv.vrf.Verify(results, sprog)
+	delete(srv.results, pi)
+	delete(srv.left, pi)
+	delete(srv.serializedProgs, pi)
+}
+
+// cleanup is called when a vm.Instance crashes.
+func (srv *RPCServer) cleanup(vmIdx int) {
+	srv.mu.Lock()
+	defer srv.mu.Unlock()
+
+	// Find all the Runners executing on the VM that crashed
+	for rnName := range srv.runners {
+		if !runOnVM(fmt.Sprintf("%d", vmIdx), rnName) {
```

Comment:
I think this can be simpler if we create the name from vmIdx the same way we do when we start a VM, and then we can just use the name directly and we don't need to loop over all srv.runners, nor runOnVM, and it seems that srv.runners itself won't be needed (at least in it's current form).  would store all pending programs for a runner directly, then it should simplify this function further as we could just loop over all pending programs directly along the lines of:
`for prog := range srv.runners[formatVMName(idx)] { ... update prog ... }`